### PR TITLE
Add Wayfinder for navigating relationships.

### LINKED ISSUE #641: Extract query_service-dependent methods from decorators into `Wayfinder`
* Reduce multiple responsibility in one class 
* Hopefully will help with #518

### PR DESCRIPTION
**Pull request by tpendragon**
Closes #641

### CODE REVIEW

File: app/decorators/ephemera_box_decorator.rb [hunk 1]
```diff
@@ -12,28 +12,29 @@ def title
...
+    wayfinder.decorated_ephemera_projects
   end
 
   def collection_slugs
     @collection_slugs ||= ephemera_projects.map(&:slug)
   end
 
   def ephemera_project
-    @ephemera_box ||= query_service.find_parents(resource: model).to_a.first.try(:decorate) || NullProject.new
+    ephemera_projects.first || NullProject.new
```
[the remark this answers is not in the document]
Comment (tpendragon) [reply app/decorators/ephemera_box_decorator.rb]: I think it depends. In this case I'm not making that decision, just matching what existed.

File: app/decorators/ephemera_folder_decorator.rb [hunk 2]
```diff
@@ -214,6 +210,7 @@ def find_resource(resource_id)
       resource_id
     end
 
+    # Unsure if I should move this to wayfinder.
```
Comment (escowles) [inline app/decorators/ephemera_folder_decorator.rb]: I think this relationship is a little different from the others.  But I do expect we'll have more use of controlled vocabs in the future, so finding a good pattern for handling these seems good.

File: spec/decorators/scanned_resource_decorator_spec.rb [hunk 3]
```diff
@@ -67,4 +67,47 @@
...
+      expect(members.map(&:id)).to eq [member.id]
+      expect(members.map(&:class)).to eq [ScannedResource]
+    end
+  end
+
+  describe "#volumes" do
+    let(:resource) { FactoryBot.create_for_repository(:scanned_resource, member_ids: [member.id, member2.id]) }
+    let(:member) { FactoryBot.create_for_repository(:scanned_resource) }
+    let(:member2) { FactoryBot.create_for_repository(:file_set) }
+    it "returns all ScannedResource members" do
```
Comment (hackmastera) [inline spec/decorators/scanned_resource_decorator_spec.rb]: it "returns all ScannedResource members, decorated"

File: spec/decorators/scanned_resource_decorator_spec.rb [hunk 4]
```diff
@@ -67,4 +67,47 @@
...
+      expect(volumes.map(&:id)).to eq [member.id]
+      expect(volumes.map(&:class)).to eq [described_class]
+    end
+  end
+
+  describe "#file_sets" do
+    let(:resource) { FactoryBot.create_for_repository(:scanned_resource, member_ids: [member.id, member2.id]) }
+    let(:member) { FactoryBot.create_for_repository(:scanned_resource) }
+    let(:member2) { FactoryBot.create_for_repository(:file_set) }
+    it "returns all FileSet members" do
```
Comment (hackmastera) [inline spec/decorators/scanned_resource_decorator_spec.rb]: "..., decorated"

File: spec/wayfinders/wayfinder_spec.rb [hunk 5]
```diff
@@ -0,0 +1,703 @@
...
+
+        wayfinder = described_class.for(resource)
+
+        expect(wayfinder.members.map(&:id)).to eq [member.id]
+        expect(wayfinder.members.map(&:class)).to eq [ScannedResource]
+      end
+    end
+
+    describe "#scanned_resources" do
+      it "returns undecorated scanned_resources" do
```
Comment (hackmastera) [inline spec/wayfinders/wayfinder_spec.rb]: "returns undecorated scanned_resource members"

File: spec/wayfinders/wayfinder_spec.rb [hunk 6]
```diff
@@ -0,0 +1,703 @@
...
+
+        wayfinder = described_class.for(resource)
+
+        expect(wayfinder.scanned_resources.map(&:id)).to eq [member.id]
+        expect(wayfinder.scanned_resources.map(&:class)).to eq [ScannedResource]
+      end
+    end
+
+    describe "#decorated_scanned_resources" do
+      it "returns decorated scanned_resources" do
```
Comment (hackmastera) [inline spec/wayfinders/wayfinder_spec.rb]: "returns decorated scanned_resource members"

File: spec/wayfinders/wayfinder_spec.rb [hunk 7]
```diff
@@ -0,0 +1,703 @@
...
+
+        wayfinder = described_class.for(resource)
+
+        expect(wayfinder.decorated_scanned_resources.map(&:id)).to eq [member.id]
+        expect(wayfinder.decorated_scanned_resources.map(&:class)).to eq [ScannedResourceDecorator]
+      end
+    end
+
+    describe "#ephemera_folders" do
+      it "returns undecorated ephemera_folders" do
```
Comment (hackmastera) [inline spec/wayfinders/wayfinder_spec.rb]: "returns undecorated ephemera_folder members"

File: spec/wayfinders/wayfinder_spec.rb [hunk 8]
```diff
@@ -0,0 +1,703 @@
...
+
+        wayfinder = described_class.for(resource)
+
+        expect(wayfinder.ephemera_folders.map(&:id)).to eq [member.id]
+        expect(wayfinder.ephemera_folders.map(&:class)).to eq [EphemeraFolder]
+      end
+    end
+
+    describe "#decorated_ephemera_folders" do
+      it "returns decorated ephemera_folders" do
```
Comment (hackmastera) [inline spec/wayfinders/wayfinder_spec.rb]: "returns decorated ephemera_folder members"

File: spec/wayfinders/wayfinder_spec.rb [hunk 9]
```diff
@@ -0,0 +1,703 @@
...
+      it "returns the ephemera box for an ephemera folder in a box" do
+        folder = FactoryBot.create_for_repository(:ephemera_folder)
+        box = FactoryBot.create_for_repository(:ephemera_box, member_ids: folder.id)
+
+        wayfinder = described_class.for(folder)
+
+        expect(wayfinder.ephemera_box.id).to eq box.id
+        expect(wayfinder.ephemera_box.class).to eq EphemeraBox
+      end
+    end
```
Comment (hackmastera) [inline spec/wayfinders/wayfinder_spec.rb]: Add a test for a folder not in a box?

File: spec/wayfinders/wayfinder_spec.rb [hunk 10]
```diff
@@ -0,0 +1,703 @@
...
+        expect(wayfinder.ephemera_boxes.map(&:id)).to eq [box.id]
+        expect(wayfinder.ephemera_boxes.map(&:class)).to eq [EphemeraBox]
+        expect(wayfinder.query_service).to have_received(:find_members).with(resource: project, model: EphemeraBox)
+
+        expect(wayfinder.decorated_ephemera_boxes.map(&:id)).to eq [box.id]
+        expect(wayfinder.decorated_ephemera_boxes.map(&:class)).to eq [EphemeraBoxDecorator]
+      end
+    end
+
+    describe "#ephemera_folders" do
```
Comment (hackmastera) [inline spec/wayfinders/wayfinder_spec.rb]: We need this and the ones on line 42/55?

File: spec/wayfinders/wayfinder_spec.rb [hunk 11]
```diff
@@ -0,0 +1,703 @@
...
+        expect(wayfinder.ephemera_fields.map(&:id)).to eq [field.id]
+        expect(wayfinder.ephemera_fields.map(&:class)).to eq [EphemeraField]
+        expect(wayfinder.query_service).to have_received(:find_members).with(resource: project, model: EphemeraField)
+
+        expect(wayfinder.decorated_ephemera_fields.map(&:id)).to eq [field.id]
+        expect(wayfinder.decorated_ephemera_fields.map(&:class)).to eq [EphemeraFieldDecorator]
+      end
+    end
+
+    describe "#templates" do
```
Comment (hackmastera) [inline spec/wayfinders/wayfinder_spec.rb]: dup of 106?

File: spec/wayfinders/wayfinder_spec.rb [hunk 12]
```diff
@@ -0,0 +1,703 @@
+# frozen_string_literal: true
+require "rails_helper"
+
+RSpec.describe Wayfinder do
+  context "when given a ScannedResource" do
```
Comment (hackmastera) [inline spec/wayfinders/wayfinder_spec.rb]: this context section is quite large; it seems to encompass more than just scanned resources stuff.

File: app/wayfinders/base_wayfinder.rb [hunk 13]
```diff
@@ -0,0 +1,82 @@
+# frozen_string_literal: true
+class BaseWayfinder
```
Comment (hackmastera) [inline app/wayfinders/base_wayfinder.rb]: Can you add class level docs about the intended use of this class? e.g. use in decorators. Also we had talked about using in controllers. It seems like that would still work; maybe a little something about what that would look like would fit here.

File: app/wayfinders/wayfinder.rb [hunk 14]
```diff
@@ -0,0 +1,31 @@
+# frozen_string_literal: true
+class Wayfinder < BaseWayfinder
```
Comment (hackmastera) [inline app/wayfinders/wayfinder.rb]: Please add class doc noting that this is the wayfinder factory slash default wayfinder for anything that doesn't define its own

File: spec/wayfinders/wayfinder_spec.rb (hunk 10 above)
Comment (tpendragon) [reply spec/wayfinders/wayfinder_spec.rb]: Yeah - I reorganized, they're for methods that need to exist for different contexts, project vs box. I fixed up the contexts to show that better.

**Pull request merged**